dynamic-libraries.md

```diff
@@ -0,0 +1,431 @@
+
+Libraries are usually not searched for in the current directory on more modern Linux-based systems. The paths which Python searches for libraries are usually not set to the current working directory either. Let's set them this way for convenience.
+
+~~~bash
+export LD_LIBRARY_PATH=.:
+export PATH=.:$PATH
+~~~
+
+Now we can create a file `helloworld.lisp` with following content:
+
+~~~lisp
+(require '#:asdf)
+(asdf:load-system :sbcl-librarian)
+
+(defpackage libhelloworld
+  (:use :cl :sbcl-librarian))
+
+(in-package libhelloworld)
+
+;; will be called from Python
+(defun hello-world (a b)
+  (+ a b))
+
+
+;; error enum to be used in C/Python code for error handling
+(define-enum-type error-type "err_t"
+  ("ERR_SUCCESS" 0)
+  ("ERR_FAIL" 1))
+
+;; mapping Common Lisp conditions to C enums
+;; in this simple example, all conditions are mapped to number 1
+;; which is "ERR_FAIL" in `error-type` enum
+(define-error-map error-map error-type 0
+  ((t (lambda (condition)
+        (declare (ignore condition))
+        (return-from error-map 1)))))
+
+;; structure of the generated C source file
+(define-api libhelloworld-api (:error-map error-map              ; error enum
+                               :function-prefix "helloworld_")   ; prefix for all function names (C doesn't have namespaces)
+  (:literal "/* types */")        ; just a comment (whatever is there will be printed as-is)
+  (:type error-type)              ; outputs the error enum
+  (:literal "/* functions */")
+  (:function                      ; function declaration - name, return type, argument types
+     (hello-world :int ((a :int) (b :int)))))
+
+;; definition of the whole library - what is there
+(define-aggregate-library libhelloworld (:function-linkage "LIBHELLOWORLD_API")
+  sbcl-librarian:handles sbcl-librarian:environment libhelloworld-api)
+
+;; builds the bindings
+(build-bindings libhelloworld ".")
+(build-python-bindings libhelloworld ".")
+
+;; outputs the Lisp core
+(build-core-and-die libhelloworld "." :compression t)
+~~~
+
+The macro `define-enum-type` creates a mapping between conditions signaled by Common Lisp functions and a return type for the wrapping C functions. If a condition is signaled from Common Lisp, it is translated into a number — a C function return value — within `define-error-map`. The enumeration type adds a C `enum`, so instead of:
+
+~~~C
+if (1 == cl_function()) {
+~~~
+
+you can write:
+
+~~~C
+if (ERR_FAIL == cl_function()) {
+~~~
+
+which is more readable.
+
+`define-api` outlines the structure of the library code to be created, specifying the error map, types, functions, and their order (`:literal` is used for comments in this case). The function `call-callback` uses previously defined types for its arguments: the `callback` type for the first argument named `fn` and the `:char-buffer` type for its second argument `out_buffer`. The `:function-prefix` option means the actual name of the exported function will be `callback_call_callback`.
```

Comment:
"call-callback" and "char-buffer" are used later and shouldn't appear in this sentence at this point?

Reply:
Sorry about that, "copy-pasta" :) I moved the article to an earlier subchapter and forgot to remove the sentence.